### PR TITLE
Add NetworkBoundResource class

### PR DESCRIPTION
**Pull request by jackstenglein**
Adds the NetworkBoundResource, Resource and ApiResponse classes in order to support fetching from the server and caching in Room. These classes are based off the GithubBrowser architecture component sample project. That project's NetworkBoundResource can be found [here](https://github.com/android/architecture-components-samples/blob/b1a194c1ae267258cd002e2e1c102df7180be473/GithubBrowserSample/app/src/main/java/com/android/example/github/repository/NetworkBoundResource.java).

### CODE REVIEW

File: MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    private static final String LOCAL_VALUE = "Local Value";
+    private static final String SERVER_VALUE = "Server Value";
+
+    @Rule
+    public InstantTaskExecutorRule instantExecutorRule = new InstantTaskExecutorRule();
+
+    private Object[] invokesCorrectMethodsParameters() {
+        return new Object[] {
+            new Object[] {false, new boolean[] {false, true, true, false} },
```
Comment (TianyiMu) [inline MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:27]: I am a bit confused here. When shouldFetch=false, why is the shouldFetch called?

File: MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            @Override protected LiveData<String> loadFromRoom() {
+                return new MutableLiveData<>(LOCAL_VALUE);
+            }
+
+            @Override protected LiveData<ApiResponse<String>> performNetworkCall() {
+                return new MutableLiveData<>(new ApiResponse(Response.success(SERVER_VALUE)));
+            }
+        };
+
+        resource.getResult().observeForever(data -> {});
```
Comment (TianyiMu) [inline MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:84]: Is this test missing assert statements?

File: MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java (hunk 1 above)
Comment (jackstenglein) [reply MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:27]: Perhaps I can rename the variables a bit better. The first parameter is the value returned by shouldFetch

File: MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java (hunk 2 above)
Comment (jackstenglein) [reply MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:84]: No, the asserts are actually in the NetworkBoundResource methods for this tests. I'll move them so the test is more clear.

File: MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java (hunk 1 above)
Comment (TianyiMu) [reply MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:27]: Ah I see. Thanks

File: MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+        assertThat(invocations).isEqualTo(expectedInvocations);
+    }
+
+    @Test
+    public void networkBoundResource_passesCorrectParameters() {
+        final String[] shouldFetchParameter = new String[1];
+        final String[] saveCallResultParameter = new String[1];
```
Comment (TianyiMu) [inline MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:66]: Why not use String?

File: MoviesTVSentiments/app/src/main/java/com/google/moviestvsentiments/service/web/Resource.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
+package com.google.moviestvsentiments.service.web;
+
+/**
+ * A container for a value, its loading status and an error message.
+ * @param <T> The type of value stored.
+ */
+public class Resource<T> {
```
Comment (TianyiMu) [inline MoviesTVSentiments/app/src/main/java/com/google/moviestvsentiments/service/web/Resource.java:7]: Wondering if it is possible to use autovalue for this?

Comment (TianyiMu) [reply MoviesTVSentiments/app/src/main/java/com/google/moviestvsentiments/service/web/Resource.java:7]: Also does it need equals? https://drive.google.com/file/d/1mW8CaRfZmeTu_lnf3EdBStCntVpqcnSB/view?usp=sharing

File: MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java (hunk 3 above)
Comment (jackstenglein) [reply MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:66]: That results in the error: `Variable 'shouldFetchParameter' is accessed from within inner class, needs to be final or effectively final`

Comment (TianyiMu) [reply MoviesTVSentiments/app/src/test/java/com/google/moviestvsentiments/service/web/NetworkBoundResourceTest.java:66]: I see. thanks

**Pull request merged**